### ISSUE TITLE
saveToKafka result Record#commit semantics

### ISSUE THREAD

**Issue by richardhundt** [cjdev/kafka-rx#12]
Hi, awesome project!

There's something I'm trying to grok though. In the README there's an example on producing to Kafka:

```Scala
tweetStream.map(parse)
  .groupBy(hashtag)
  .foreach { (tag, subStream) =>
    subStream.map(toProducerRecord)
      .saveToKafka(kafkaProducer)
      .foreach { savedRecord =>
        savedRecord.commit() // checkpoint position in the source stream
      }
  }
```

The `savedRecord` is synthesised and an `Observable[Record]` is returned. What's puzzling is the `savedRecord.commit()` call. Am I reading that right in that it is _not_ committing to `savedRecord.offset`, but instead the `Record` produced by `subStream`?

So to put it another way, is the `commitfn` attached to `savedRecord` a closure torn off and bound to the incoming record, and that's the offset getting committed?

I'm finding it a little tricky to unpack the offset management, so any pointers here would be great.
Thanks!

**Comment by eggsby** [comment]
Hi Richard,

Glad to hear that you are interested in the project!

This example is likely unclear since there is no explanation of what `toProducerRecord` does.

Each `Record` has a `produce` method which lets you `derive` a new record (specifically a new kafka.clients `ProducerRecord`) while maintaining the `commit` context.

You are exactly right that each record gets its commit context is chunked off as a closure (that happens [here](https://github.com/cjdev/kafka-rx/blob/master/src/main/scala/com/cj/kafka/rx/OffsetManager.scala#L33-L37)) -- then that method reference is passed around as derivative values are made.

It's worth mentioning that this behavior (specifically the threading of this commit closure when using `saveToKafka`) is broken in 0.3.0 due to a regression that got [resolved](9606161b17e6e39d24fceb8e95d6f5e6db08b295) in 0.3.1.

I'll push 0.3.1 to maven central ASAP to get this fix out there :)

Cheers, let me know if this helps.

-- Thomas

**Comment by richardhundt** [comment]
Hi Thomas, thanks very much for the explanation. Makes sense, and I look forward to the fix!
